File: game-server/src/main/java/com/osroyale/game/event/impl/NpcClickEvent.java
package com.osroyale.game.event.impl;

import com.osroyale.game.event.Event;
import com.osroyale.game.world.entity.mob.npc.Npc;

public class NpcClickEvent implements Event {

    private final int type;

    private final Npc npc;

    public NpcClickEvent(int type, Npc npc) {
        this.type = type;
        this.npc = npc;
    }

    public int getType() {
        return type;
    }

    public Npc getNpc() {
        return npc;
    }

}

File: game-server/src/main/kotlin/org/jire/tarnishps/event/npc/NpcClickEvent.kt
package org.jire.tarnishps.event.npc

import com.osroyale.game.world.World
import com.osroyale.game.world.entity.mob.data.PacketType
import com.osroyale.game.world.entity.mob.npc.Npc
import com.osroyale.game.world.entity.mob.player.Player
import kotlin.jvm.optionals.getOrNull

/**
 * @author Jire
 */
interface NpcClickEvent : NpcEvent {

    val slot: Int

    override fun canHandle(player: Player) = super.canHandle(player)
            && !player.locking.locked(PacketType.CLICK_NPC)

    override fun handle(player: Player) {
        val npc = World.getNpcBySlot(slot).getOrNull() ?: return
        if (!npc.isValid) return

        val position = npc.position
        val region = World.getRegions().getRegion(position)
        if (!region.containsNpc(position.height, npc)) return

        handleNpc(player, npc)
    }

    fun handleNpc(player: Player, npc: Npc) {}

}